src/main.rs

```diff
@@ -141,9 +152,32 @@ fn main() {
             arch = built_info::CFG_TARGET_ARCH
         );
 
-        let (mut topology, mut graceful_crash) =
-            topology::start(config, &mut rt, opts.require_healthy)
-                .unwrap_or_else(|| std::process::exit(1));
+        let (exit_after, require_healthy) = if opts.dry_run {
+            info!("Dry run enabled, exiting after config validation");
+            if opts.healthchecks_only {
+                info!("Ignoring --healthchecks-only due to more strong option --dry-run");
+            }
+
+            (Some(Stage::ConfigValidated), opts.require_healthy)
+        } else if opts.healthchecks_only {
+            info!("Exit after finishing healthchecks configured");
+            (Some(Stage::HealthChecksPassed), true)
+        } else {
+            (None, opts.require_healthy)
+        };
+
+        let result = topology::start_or_validate(config, &mut rt, exit_after, require_healthy);
+        result.as_ref().left().map(|success| {
+            let exit_code = if *success {
+                exitcode::OK
+            } else {
+                exitcode::CONFIG
+            };
+
+            std::process::exit(exit_code);
+        });
+
+        let (mut topology, mut graceful_crash) = result.right().unwrap();
```

Comment:
what are the cases taht this unwrap will fire?

Reply:
Actually, it can't fire because we just handled `left()` case with `std::process::exit`.

For clarity, better to rewrite this `left()` and `right()` cases with pattern-matching, but this would affect all lines below in this method, not sure that it is good to touch lines without necessity..

P.S. I wish this `either` crate had method `unwrap_or`, so I could join both handlers in one call. But `either` in Rust seems to be side-unbiased, that's probably why they haven't added this method.

Reply:
This is an illustration of what I mean, made with `Result` type. This could be a bit confusing, because left side is considered to be an error and not a status which can be legal exit :) But probably  I'm getting too peaky about it. The code itself looks pretty elegant for me:

https://github.com/timberio/vector/commit/92775af04c0f9d471ed82a5cca0ccea6b7179cea

Reply:
Oh, ignore it, I found the method I needed in `either` crate: it's called `right_or_else`.
I made changes to code with it, now branches handling is clearer.